TrainTest fails sometimes

======================================================================
FAIL: test_ppo_cart_pole (bin.train_test.TrainTest)
test_ppo_cart_pole (bin.train_test.TrainTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/ALF/alf/bin/train_test.py", line 100, in test_ppo_cart_pole
    self._test_train('ppo_cart_pole.gin', _test_func)
  File "/ALF/alf/bin/train_test.py", line 126, in _test_train
    assert_func(episode_returns, episode_lengths)
  File "/ALF/alf/bin/train_test.py", line 97, in _test_func
    self.assertGreater(np.mean(returns[-2:]), 198)
AssertionError: 197.8499984741211 not greater than 198

@witwolf It seems that the determinism isn't working as expected? There are some other cases failed sometimes, and I have to manually restart the testing job each time.

--- a/alf/environments/utils.py
+++ b/alf/environments/utils.py
@@ -17,6 +17,7 @@ import traceback
 import tensorflow as tf
 import gin.tf
 from absl import logging
+import numpy as np
 
 from tf_agents.environments import suite_gym, parallel_py_environment, tf_py_environment
 
@@ -155,9 +156,16 @@ def create_environment(env_name='CartPole-v0',
     if nonparallel:
         # Each time we can only create one unwrapped env at most
         py_env = env_load_fn(env_name)
+        py_env.seed(np.random.randint(0, np.iinfo(np.int32).max))
     else:
         py_env = parallel_py_environment.ParallelPyEnvironment(
             [lambda: env_load_fn(env_name)] * num_parallel_environments)
+
+        py_env.seed([
+            np.random.randint(0,
+                              np.iinfo(np.int32).max)
+            for i in range(num_parallel_environments)
+        ])
 
     return tf_py_environment.TFPyEnvironment(py_env)
 

